File: Baekjoon/유석환/Q1149.swift
typealias Cost = (R: Int, G: Int, B: Int)

let N = Int(readLine()!)!
var cost = [Cost](repeating: Cost(0, 0, 0), count: 1001)

for i in 1...N {
    let rgb = readLine()!.split(separator: " ").map { Int(String($0))! }
    cost[i] = Cost(rgb[0], rgb[1], rgb[2])
}

var dpTable = [Cost](repeating: Cost(0, 0, 0), count: 1001)
for i in 1...N {
    dpTable[i].R = min(dpTable[i - 1].G, dpTable[i - 1].B) + cost[i].R
    dpTable[i].G = min(dpTable[i - 1].R, dpTable[i - 1].B) + cost[i].G
    dpTable[i].B = min(dpTable[i - 1].R, dpTable[i - 1].G) + cost[i].B
}

print(min(dpTable[N].R, dpTable[N].G, dpTable[N].B))

File: Baekjoon/정우빈/Q1149.kt
import kotlin.math.min
// 각 집의 색 비용
lateinit var q1149CostArr: Array<Array<Int>>
// 누적 비용
lateinit var q1149Arr: Array<Array<Int>>
fun main() {
    val br = System.`in`.bufferedReader()
    val bw = System.`out`.bufferedWriter()

    val count = br.readLine().trim().toInt()
    q1149CostArr = Array(count) { Array(3) { 0 } }
    q1149Arr = Array(count) { Array(3) { 0 } }

    // 각 집의 색 비용 초기화
    for (i in 0 until count) {
        val (a, b, c) = br.readLine().trim().split(" ").map { it.toInt() }
        // 0 : red, 1 : green, 2 : blue
        q1149CostArr[i][0] = a
        q1149CostArr[i][1] = b
        q1149CostArr[i][2] = c
    }

    // 첫번째 집 누적 비용 초기화
    q1149Arr[0][0] = q1149CostArr[0][0]
    q1149Arr[0][1] = q1149CostArr[0][1]
    q1149Arr[0][2] = q1149CostArr[0][2]

    // 3가지 색 중 최소비용 출력
    bw.write("${min(q1149Func(0, count - 1), min(q1149Func(1, count - 1), q1149Func(2, count - 1)))}")
    bw.flush()
    bw.close()
    br.close()
}

// 마지막 집의 색의 최솟값 구하는 함수
fun q1149Func(color: Int, index: Int): Int {
    if (q1149Arr[index][color] == 0) {
        when(color) {
            0 -> q1149Arr[index][color] = min(q1149Func(1, index - 1), q1149Func(2, index - 1)) + q1149CostArr[index][color]
            1 -> q1149Arr[index][color] = min(q1149Func(0, index - 1), q1149Func(2, index - 1)) + q1149CostArr[index][color]
            2 -> q1149Arr[index][color] = min(q1149Func(0, index - 1), q1149Func(1, index - 1)) + q1149CostArr[index][color]
        }
    }
    return q1149Arr[index][color]
}